[WIP] Codegen: Work towards supporting genex

xsacha opened the pull request:
MSVC currently fails compiling with Release;Debug due to using "/OX" flag with Debug mode. This is because the check for CMAKE_BUILD_TYPE is flawed (it is empty).
This is a step towards fixing this issue the proper way by using generator expressions.
generator expressions on set_source_file_properties weren't added until around CMake 3.7 but set_target is supported earlier

I want to use `$<$<NOT:$<CONFIG:Debug>>:/Ox>` to set Ox but this will currently fail later down the code where set_source_files_properties is used.

xsacha commented on cmake/Codegen.cmake:
```diff
@@ -53,35 +53,35 @@ install(FILES ${CMAKE_BINARY_DIR}/caffe2/core/macros.h
-  ENDIF()
-  SET(VCOMP_LIB "vcomp")
+  if(MSVC)
+    set(OPT_FLAG "/0X /fp:strict ")
```
Thanks!

peterjc123 commented on cmake/Codegen.cmake:
```diff
@@ -53,35 +53,35 @@ install(FILES ${CMAKE_BINARY_DIR}/caffe2/core/macros.h
+  else()
+    set(OPT_FLAG "-O3 ")
+  endif()
+  SET(VCOMP_LIB vcomp$<$<CONFIG:Debug>:d>)
```
This one is not expanded as I can see this line in the log.
```
07:10:41 cmd.exe /C "cd . && "C:\Program Files\CMake\bin\cmake.exe" -E vs_link_dll --intdir=caffe2\CMakeFiles\caffe2.dir --manifests  -- C:\PROGRA~2\MIB055~1\2017\COMMUN~1\VC\Tools\MSVC\1411~1.255\bin\Hostx64\x64\link.exe /nologo @CMakeFiles\caffe2.rsp  /out:bin\caffe2.dll /implib:LIB\caffe2.lib /pdb:bin\caffe2.pdb /dll /version:0.0 /INCREMENTAL:NO /NODEFAULTLIB:vcomp$<$<CONFIG:Debug>:d>   && cd ."
07:10:41 && was unexpected at this time.
```

xsacha replied:
Thanks.
I wanted to avoid all the nasty if/else checking everywhere but seems like I'll have to do it.

Is it possible at all to bump the CMake minimum version or is that already new enough (3.5)?

peterjc123 replied:
I'm afraid this is not the issue here. We are using CMake 3.12.

peterjc123 commented on cmake/Codegen.cmake:
```diff
@@ -53,35 +53,35 @@ install(FILES ${CMAKE_BINARY_DIR}/caffe2/core/macros.h
   IF(C_AVX_FOUND)
     IF(MSVC)
-      SET_SOURCE_FILES_PROPERTIES(${CMAKE_CURRENT_LIST_DIR}/../aten/src/TH/vector/AVX.cpp PROPERTIES COMPILE_FLAGS "${OPT_FLAG}/arch:AVX ${CXX_AVX_FLAGS}")
+      set_property(SOURCE ${CMAKE_CURRENT_LIST_DIR}/../aten/src/TH/vector/AVX.cpp APPEND PROPERTY CMAKE_CXX_FLAGS ${OPT_FLAG} /arch:AVX ${CXX_AVX_FLAGS})
```
And those changes are causing the build failures on Linux.

xsacha replied:
Thanks. I'll get build working on Linux and then send again.

Commit:
Codegen: Work towards supporting genex

        MSVC currently fails compiling with Release;Debug due to using /OX flag with Debug mode. This is because the check for CMAKE_BUILD_TYPE is flawed (it is empty).
        This is a step towards fixing this issue the proper way by using generator expressions.
        generator expressions on set_source_file_properties weren't added until around CMake 3.7 but set_target is supported earlier

        I want to use $<$<NOT:$<CONFIG:Debug>>:/Ox> to set Ox but this will currently fail later down the code where set_source_files_properties is used.

ezyang commented:
Remove the WIP mark when Linux is passing, thanks!
